variations/build-variations.mjs

```diff
@@ -0,0 +1,73 @@
+import fs from 'fs-extra';
+import fsorig from 'fs';
+import deepmerge from 'deepmerge';
+import { dirname } from 'path';
+import { fileURLToPath } from 'url';
+
+const localpath = dirname( fileURLToPath( import.meta.url ) );
+
+(async function start() {
+	let args = process.argv.slice(2);
+	let source = args?.[0];
+	let variation = args?.[1];
+	if ( source && variation ) {
+		return await buildVariation(source, variation);
+	}
+	return await buildAllVariations();
+})();
+
+async function buildAllVariations(){
+	for (let source of getDirectories( localpath )){
+		for (let variation of getDirectories( `${localpath}/${source}` )){
+			await buildVariation(source, variation);
+		}
+	}
+}
+
+async function buildVariation(source, variation) {
+	const srcDir = localpath + '/../' + source;
+	const srcVariationDir = localpath + '/' + source + '/' + variation;
+	const destDir = localpath + '/../' + variation;
+
+	console.log( `Copying the source ${source} to the variation ${variation}` );
+
+	try {
+
+		// First empty the old directory.
+		await fs.emptyDir( destDir );
+
+		// Then copy the source directory.
+		await fs.copy( srcDir, destDir );
+
+		// remove unneeded resources
+		await fs.remove( destDir + '/sass');
+		await fs.remove( destDir + '/node_modules' );
+
+		// copy the variation directory.
+		await fs.copy( srcVariationDir, destDir );
+
+		// copy the readme
+		await fs.copy( localpath + '/variation-readme.md', destDir + '/variation-readme.md' );
+
+		// merge the theme.json files
+		const srcJsonFile = await fs.readFile( srcDir + '/theme.json', 'utf8' );
+		const srcVariationJsonFile = await fs.readFile( srcVariationDir + '/theme.json', 'utf8' );
+		const srcJson = JSON.parse( srcJsonFile );
+		const srcVariationJson = JSON.parse( srcVariationJsonFile );
+		const mergedJson = deepmerge(srcJson, srcVariationJson, {
```

Comment:
My concern with this is that this isn't how Gutenberg merges theme.json files, so we might get different results when we try to port this to a child theme.

Reply:
What does it do differently?  I thought that's how it did it.  Merge objects, replace arrays.

Reply:
Oh you might be right. The way Gutenberg does it uses PHP and it's not very smart, so I know we need to be careful to copy it correctly. I'll investigate...

Reply:
[This](https://github.com/WordPress/gutenberg/blob/12.0.01/lib/class-wp-theme-json-gutenberg.php#L1297) is where that logic was for v12.

But I can't find the same in [trunk](https://github.com/WordPress/gutenberg/tree/trunk/lib)!

Reply:
Do we need to note the limits to things that can be customized with our simple approach?  I think those basic rules (merge objects, replace arrays) get us the basic functionality we need for variations.

Reply:
Could we not reuse the `merge` function from Gutenberg itself?

Reply:
I can't see any way to do that. I think for now this approach is fine. If we notice issues we can iterate :)